### PR TITLE
Upload release binaries to new release infrastructure

### PR DESCRIPTION
**Pull request by justinas**
Tag pipeline: https://drone.teleport.dev/gravitational/teleport/9607

Part of https://github.com/gravitational/cloud/milestone/36

### CODE REVIEW

**Review by tcsc — APPROVED** [review]

**Comment by russjones** [comment]
@justinas We're also adding builds with the following format, will the new CDN support them as well?

```
teleport-8.0.1-linux-amd64-centos7-bin.tar.gz
teleport-ent-8.0.1-linux-amd64-centos7-bin.tar.gz
teleport-ent-8.0.1-linux-amd64-centos7-fips-bin.tar.gz
```

**Review by quinqu — APPROVED** [review]

**Comment by justinas** [comment]
> @justinas We're also adding builds with the following format, will the new CDN support them as well?
> 
> ```
> teleport-8.0.1-linux-amd64-centos7-bin.tar.gz
> teleport-ent-8.0.1-linux-amd64-centos7-bin.tar.gz
> teleport-ent-8.0.1-linux-amd64-centos7-fips-bin.tar.gz
> ```

@russjones It should support these automatically, but I will rebase and make a test tag to confirm.

File: dronegen/tag.go [hunk 1]
```diff
@@ -308,6 +319,34 @@ func tagCopyPackageArtifactCommands(b buildType, packageType string) []string {
...
+func tagCreateReleaseAssetCommands(b buildType) []string {
+	commands := []string{
+		`WORKSPACE_DIR=$${WORKSPACE_DIR:-/}`,
+		`VERSION=$(cat $WORKSPACE_DIR/go/.version.txt)`,
+		fmt.Sprintf(`RELEASES_HOST=%v`, releasesHost),
+		`echo $RELEASES_CERT | base64 -d > $WORKSPACE_DIR/releases.crt`,
+		`echo $RELEASES_KEY | base64 -d > $WORKSPACE_DIR/releases.key`,
+		`trap "rm -f /releases.crt /releases.key" EXIT`,
+		`CREDENTIALS="--cert $WORKSPACE_DIR/releases.crt --key $WORKSPACE_DIR/releases.key"`,
+		`which curl || apk add --no-cache curl`,
```
Comment (codingllama) [inline dronegen/tag.go:332]: Don't `apk add` need some sort of `-y` flag to not block on input?

File: dronegen/tag.go [hunk 2]
```diff
@@ -308,6 +319,34 @@ func tagCopyPackageArtifactCommands(b buildType, packageType string) []string {
...
+		fmt.Sprintf(`cd $WORKSPACE_DIR/go/artifacts
+for file in $(find . -type f ! -iname '*.sha256'); do
+  # Skip files that are not results of this build
+  # (e.g. tarballs from which OS packages are made)
+  [ -f "$file.sha256" ] || continue
+
+  product="$(basename "$file" | sed -E 's/(-|_)v?[0-9].*//')" # extract part before -vX.Y.Z
```
Comment (codingllama) [inline dronegen/tag.go:339]: Make sure pattern matches the end of the filename?

```suggestion
  product="$(basename "$file" | sed -E 's/(-|_)v?[0-9].*$//')" # extract part before -vX.Y.Z
```

this is still a bit iffy IMO, but I guess it's alright.

File: dronegen/tag.go [hunk 3]
```diff
@@ -308,6 +319,34 @@ func tagCopyPackageArtifactCommands(b buildType, packageType string) []string {
...
+  shasum="$(cat "$file.sha256" | cut -d ' ' -f 1)"
+  status_code=$(curl $CREDENTIALS -o $WORKSPACE_DIR/curl_out.txt -w "%%{http_code}" -F product=$product -F version=$VERSION -F notesMd="# Teleport $VERSION" -F status=draft $RELEASES_HOST/releases)
```
Comment (codingllama) [inline dronegen/tag.go:341]: Please double-quote variables inside the command.

Same for the commands below.

File: dronegen/tag.go [hunk 4]
```diff
@@ -308,6 +319,34 @@ func tagCopyPackageArtifactCommands(b buildType, packageType string) []string {
...
+  [ $status_code = 200 ] || [ $status_code = 409 ] || (echo "curl HTTP status: $status_code"; cat $WORKSPACE_DIR/curl_out.txt && exit 1)
```
Comment (codingllama) [inline dronegen/tag.go:342]: Let's make this into a proper if, there is too much going in the short-circuit chains.

File: dronegen/tag.go [hunk 5]
```diff
@@ -308,6 +319,34 @@ func tagCopyPackageArtifactCommands(b buildType, packageType string) []string {
...
+	commands := []string{
+		`WORKSPACE_DIR=$${WORKSPACE_DIR:-/}`,
+		`VERSION=$(cat $WORKSPACE_DIR/go/.version.txt)`,
+		fmt.Sprintf(`RELEASES_HOST=%v`, releasesHost),
+		`echo $RELEASES_CERT | base64 -d > $WORKSPACE_DIR/releases.crt`,
+		`echo $RELEASES_KEY | base64 -d > $WORKSPACE_DIR/releases.key`,
+		`trap "rm -f /releases.crt /releases.key" EXIT`,
+		`CREDENTIALS="--cert $WORKSPACE_DIR/releases.crt --key $WORKSPACE_DIR/releases.key"`,
+		`which curl || apk add --no-cache curl`,
+		fmt.Sprintf(`cd $WORKSPACE_DIR/go/artifacts
```
Comment (codingllama) [inline dronegen/tag.go:333]: I would be tempted to make this into a Go template so expanded variables are named and easier to spot. Up to you.

**Review by codingllama — COMMENTED** [review]
I'm not super familiar with dronegen, but overall looks reasonable to me.

File: dronegen/tag.go (hunk 1 above)
Comment (justinas) [reply dronegen/tag.go:332]: I don't think so, apk does not seem to ask for input in normal circumstances:

```console
$ docker run -it alpine:latest
/ # apk add --no-cache curl
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/community/x86_64/APKINDEX.tar.gz
(1/5) Installing ca-certificates (20191127-r5)
(2/5) Installing brotli-libs (1.0.9-r3)
(3/5) Installing nghttp2-libs (1.42.0-r1)
(4/5) Installing libcurl (7.79.1-r0)
(5/5) Installing curl (7.79.1-r0)
Executing busybox-1.32.1-r6.trigger
Executing ca-certificates-20191127-r5.trigger
OK: 8 MiB in 19 packages
```

Comment (codingllama) [reply dronegen/tag.go:332]: It looks like, if you want interactive, you have to pass `apk add -i`, so it's kind of the opposite of others. Thanks for looking it up.

**Review by codingllama — APPROVED** [review]

**Pull request merged**